src/compile/ast.rs

```diff
@@ -913,66 +913,75 @@ impl Type {
+                    .iter()
+                    .map(|f| f.tipe.default_value(type_tree))
+                    .collect();
+                TupleTree::fold_into_tuple(field_values)
+            }
+            Type::Func(..) => {
+                // the error codepoint
+                Value::CodePoint(CodePt::Null)
+            }
+            Type::Option(_) => Value::new_tuple(vec![Value::Int(Uint256::zero())]),
+            Type::Map(_, _) => Value::new_tuple(vec![
+                Value::from(0), // kvs.mini builtin_kvsNew() unsafe casts on 0
+                Value::from(0), // set size to 0
+            ]),
             Type::Array(t) => {
-                let (def, safe) = t.default_value();
-                (
-                    Value::new_tuple(vec![
-                        Value::Int(Uint256::one()),
-                        Value::Int(Uint256::one()),
-                        Value::new_tuple(vec![def]),
-                    ]),
-                    safe,
-                )
+                let fixed = Type::FixedArray(t.clone(), 1).default_value(type_tree);
+                Value::new_tuple(vec![
+                    Value::from(1), // size
+                    Value::from(1), // topstep
+                    fixed,          // array.mini builtin_arrayNew() unsafe casts this
+                ])
             }
-            Type::FixedArray(t, sz) => {
-                let (default_val, safe) = t.default_value();
-                let mut val = Value::new_tuple(vec![default_val; 8]);
-                let mut chunk_size = 1;
-                while chunk_size * TUPLE_SIZE < *sz {
-                    val = Value::new_tuple(vec![val; 8]);
-                    chunk_size *= 8;
+            Type::FixedArray(t, size) => {
+                // emulate array.mini builtin_arrayNew()
+                fn emulated_builtin(size: usize, mut base: Value) -> Value {
+                    let mut chunk = 1;
+                    while (8 * chunk < size) {
+                        chunk = 8 * chunk;
+                        base = Value::new_tuple(vec![base; 8]);
+                    }
+                    Value::new_tuple(vec![
+                        Value::from(size),  // size
+                        Value::from(chunk), // topstep
+                        Value::new_tuple(vec![base; 8]),
+                    ])
                 }
-                (val, safe)
+                emulated_builtin(*size, t.default_value(type_tree))
             }
-            Type::Struct(fields) => {
-                let mut vals = Vec::new();
-                let mut is_safe = true;
-                for field in fields {
-                    let (val, safe) = field.tipe.default_value();
-                    vals.push(val);
-                    is_safe = is_safe && safe;
-                }
-                (value_from_field_list(vals), is_safe)
+            Type::Nominal(..) => {
+                let tipe = self.rep(type_tree).unwrap_or(Type::Any);
```

Comment:
This should throw an error if the rep function fails. As it stands it would not be type safe on read, and that could lead to a silent error that's only discovered at run time.

Reply:
Nice catch, I was debugging something and forgot to change that back to a panic. I chose to panic since there really shouldn't be a way to make it far enough to codegen & linking with an unresolvable type -- and if we find a way to do that, we'd want to do this check while typechecking the AST.